sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotUseCollectionInItsOwnMethodCalls.cs

```diff
@@ -0,0 +1,182 @@
+                    if (operands != null &&
+                        EquivalenceChecker.AreEquivalent(operands.Left, operands.Right))
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(rule, operands.Left.GetLocation(),
+                            additionalLocations: new[] { operands.Right.GetLocation() },
+                            messageArgs: new[] { operands.Right.ToString(), operands.ErrorMessage }));
+                    }
+                }, SyntaxKind.InvocationExpression);
+        }
+
+        private static OperandsToCheck GetOperandsToCheckIfTrackedMethod(InvocationExpressionSyntax invocation,
+            SemanticModel model)
+        {
+            var methodSymbol = model.GetSymbolInfo(invocation).Symbol as IMethodSymbol;
+
+            bool isEnumerableMethod;
+            var message = ProcessIssueMessageFromMethod(methodSymbol, out isEnumerableMethod);
+            if (message == null)
+            {
+                return null;
+            }
+
+            if (isEnumerableMethod && methodSymbol.MethodKind == MethodKind.Ordinary)
+            {
+                return new OperandsToCheck(invocation.ArgumentList.Arguments[0].Expression,
+                    invocation.ArgumentList.Arguments[1].Expression, message);
+            }
+
+            var invokingExpression = (invocation.Expression as MemberAccessExpressionSyntax)?.Expression;
+            if (invokingExpression != null)
+            {
+                return new OperandsToCheck(invokingExpression, invocation.ArgumentList.Arguments[0].Expression,
+                    message);
+            }
+
+            return null;
+        }
+
+        private static string ProcessIssueMessageFromMethod(IMethodSymbol methodSymbol, out bool isEnumerableMethod)
+        {
+            isEnumerableMethod = true;
+
+            if (methodSymbol.IsEnumerableConcat())
+            {
+                return UnexpectedBehaviorMessage;
+            }
+
+            if (methodSymbol.IsEnumerableIntersect() ||
+                methodSymbol.IsEnumerableUnion())
+            {
+                return AlwaysSameCollectionMessage;
+            }
+
+            if (methodSymbol.IsEnumerableSequenceEqual())
+            {
+                return AlwaysTrueMessage;
+            }
+
+            if (methodSymbol.IsEnumerableExcept())
+            {
+                return AlwaysEmptyCollectionMessage;
+            }
+
+            isEnumerableMethod = false;
+            if (methodSymbol.IsListAddRange())
+            {
+                return UnexpectedBehaviorMessage;
+            }
+
+            if (IsISetUnionWith(methodSymbol))
+            {
+                return AlwaysSameCollectionMessage;
+            }
+
+            if (IsISetExceptWith(methodSymbol))
+            {
+                return AlwaysEmptyCollectionMessage;
+            }
+
+            return null;
+        }
+
+        private static bool IsISetUnionWith(IMethodSymbol methodSymbol)
```

Comment:
As said earlier I don't think we should extract methods that are used only once.